### ISSUE TITLE
sql: column names not being passed correctly

### ISSUE THREAD

**Issue by roncrdb** [cockroachdb/cockroach#36501]
**Describe the problem**

In psql if you give an alias to a column it will pass that name along: 

```SELECT DISTINCT *
FROM (
    SELECT unnest(ARRAY[439889519552299009]) AS filter_id2
) AS uq
JOIN (
    SELECT unnest(ARRAY[439889519552299009]) AS filter_id
) AS ab
ON uq.filter_id2 = ab.filter_id;
     filter_id2     |     filter_id
--------------------+--------------------
 439889519552299009 | 439889519552299009
```

In CRDB, with a single column, we do not pass the name correctly: 

```
root@:26257/test> SELECT DISTINCT *   FROM ( SELECT unnest(ARRAY[439889519552299009]) AS filter_id2, 1 AS hello) AS uq JOIN (SELECT unnest(ARRAY[439889519552299009]) AS filter_id, 2 AS goodbye) AS ab ON uq.filter_id2 = ab.filter_id;
      filter_id2     | hello |     filter_id      | goodbye
+--------------------+-------+--------------------+---------+
  439889519552299009 |     1 | 439889519552299009 |       2
(1 row)

Time: 1.993ms

root@:26257/test> SELECT DISTINCT *FROM (SELECT unnest(ARRAY[439889519552299009]) AS filter_id2, 1 AS hello) AS uq JOIN (SELECT unnest(ARRAY[439889519552299009]) AS filter_id) AS ab ON uq.filter_id2 = ab.filter_id;

pq: column "ab.filter_id" does not exist

root@:26257/test> SELECT DISTINCT * FROM ( SELECT unnest(ARRAY[439889519552299009]) AS filter_id2, 1 AS hello) AS uq JOIN (SELECT unnest(ARRAY[439889519552299009]) AS filter_id) AS abON uq.filter_id2 = ab;
      filter_id2     | hello |         ab
+--------------------+-------+--------------------+
  439889519552299009 |     1 | 439889519552299009
(1 row)

Time: 6.792ms
```
**To Reproduce**

What did you do? Describe in your own words.

If possible, provide steps to reproduce the behavior:

1. Set up CockroachDB cluster ...
2. Send SQL ... / CLI command ...
3. Look at UI / log file / client app ...

**Expected behavior**
Column names should be passed correctly in order to avoid error message that column doesn't exist. 

**Environment:**
 - CockroachDB v2.1.6

**Comment by tim-o** [comment]
Zendesk ticket #[3239](https://cockroachdb.zendesk.com/agent/tickets/3239) has been linked to this issue.

**Comment by jordanlewis** [comment]
This appears to be a regression in the CBO. Here's a reproduction:

```
[17:50]% ./cockroach demo                                                                                                                                                                                                                                                 jordan@Termato:~/repo/src/github.com/cockroachdb/cockroach on dont-save-tmp -> mine/dont-save-tmp
#
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB
# instance. Your changes will not be saved!
#
# Web UI: http://127.0.0.1:58580
#
# Server version: CockroachDB CCL v19.1.0-beta.20190318-428-g93f796457a-dirty (x86_64-apple-darwin18.2.0, built 2019/03/28 22:28:09, go1.11) (same version as client)
# Cluster ID: cb9ce44e-7752-4d3d-b630-b65b861de820
#
# Enter \? for a brief introduction.
#
root@127.0.0.1:58579/defaultdb> SELECT * FROM (                                                                                                                                                                                                                                                                                                                                 SELECT unnest(ARRAY[439889519552299009]) AS filter_id2                                                                                                                                                                                                                                                                                                                  ) AS uq                                                                                                                                                                                                                                                                                                                                                                     JOIN (                                                                                                                                                                                                                                                                                                                                                                          SELECT unnest(ARRAY[439889519552299009]) AS filter_id                                                                                                                                                                                                                                                                                                                   ) AS ab                                                                                                                                                                                                                                                                                                                                                                     ON uq.filter_id2 = ab.filter_id;
pq: column "uq.filter_id2" does not exist
root@127.0.0.1:58579/defaultdb> set optimizer=off;
SET

Time: 447µs

root@127.0.0.1:58579/defaultdb> SELECT * FROM (                                                                                                                                                                                                                                                                                                                                 SELECT unnest(ARRAY[439889519552299009]) AS filter_id2                                                                                                                                                                                                                                                                                                                  ) AS uq                                                                                                                                                                                                                                                                                                                                                                     JOIN (                                                                                                                                                                                                                                                                                                                                                                          SELECT unnest(ARRAY[439889519552299009]) AS filter_id                                                                                                                                                                                                                                                                                                                   ) AS ab                                                                                                                                                                                                                                                                                                                                                                     ON uq.filter_id2 = ab.filter_id;
      filter_id2     |     filter_id
+--------------------+--------------------+
  439889519552299009 | 439889519552299009
(1 row)

Time: 1.516ms

root@127.0.0.1:58579/defaultdb>
```

**Comment by jordanlewis** [comment]
@RaduBerinde could you further triage?

**Comment by RaduBerinde** [comment]
I'll take a first look. Here's a smaller repro:
```
> SELECT * FROM (SELECT unnest(ARRAY[1]) AS lol) AS t WHERE t.lol = 1;
pq: column "t.lol" does not exist
```

Also:
```
root@127.70.228.24:38339/defaultdb> SELECT * FROM (SELECT unnest(ARRAY[1]) AS colalias);
  colalias  
+----------+
         1  
(1 row)

root@127.70.228.24:38339/defaultdb> SELECT * FROM (SELECT unnest(ARRAY[1]) AS colalias) AS tablealias;
  tablealias  
+------------+
           1  
(1 row)
```
Note how the column changed name when it shouldn't have.

I believe the problematic code is here:
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/optbuilder/select.go#L212

We should only be doing this if the column didn't have an alias. Not entirely sure how to make that distinction at this point though - maybe add a flag to `scopeColumn`? CC @rytaft in case you have a better idea.

**Comment by RaduBerinde** [comment]
Actually, we shouldn't be renaming it in either case. This is with postgres:
```
radu=> SELECT * FROM (SELECT unnest(ARRAY[1])) AS tablealias;
 unnest 
--------
      1
(1 row)
```

**Comment by rytaft** [comment]
Yea, I think I just copied the logic from the heuristic planner and didn't check what Postgres was doing. Although it sounds like I might have made an error in copying from the heuristic planner, too. Anyway, I think we should just get rid of those lines that you found.

Thanks for investigating!

**Comment by RaduBerinde** [comment]
A workaround is to move the column alias with the table alias:

```
SELECT * FROM
  (SELECT unnest(ARRAY[439889519552299009])) AS uq (filter_id2)
JOIN
  (SELECT unnest(ARRAY[439889519552299009])) AS ab (filter_id)
ON uq.filter_id2 = ab.filter_id
```

**Comment by jonwalch** [comment]
Huge props for moving so fast on this. When can we expect this to make it into a release?

**Comment by RaduBerinde** [comment]
It will be fixed in 19.1 which is coming very soon. Would you need it in a 2.1.x update as well or do you plan to upgrade to 19.1? (the next 2.1.x will likely be out after 19.1)

**Comment by jonwalch** [comment]
@RaduBerinde whatever works best for you, we'll use the workaround in the meantime
